# Building a Multilayer Neural Network to Classify Sonar Signals
***

## Table of contents
1. [Import and data description](#Import-and-data-description)
2. [Data preparation](#Data-preparation)
3. [Building the multilayer model](#Building-the-multilayer-model)
4. [Training configuration](#Training-configuration)
5. [Model training and evaluation](#Model-training-and-evaluation)
6. [Final remarks](#Final-remarks)

### Import and data description

[Return to top](#Building-a-Multilayer-Neural-Network-to-Classify-Sonar-Signals)  

In [1]:
#For data analysis/manipulation
import tensorflow as tf
import pandas as pd
from sklearn.preprocessing import LabelEncoder

#For Neural Network setup
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import Dropout

In [2]:
#Load sonar.csv
df = pd.read_csv('sonar.csv')
df.head()

attribute_1  attribute_2  attribute_3  attribute_4  attribute_5  \
0       0.0200       0.0371       0.0428       0.0207       0.0954   
1       0.0453       0.0523       0.0843       0.0689       0.1183   
2       0.0262       0.0582       0.1099       0.1083       0.0974   
3       0.0100       0.0171       0.0623       0.0205       0.0205   
4       0.0762       0.0666       0.0481       0.0394       0.0590   

   attribute_6  attribute_7  attribute_8  attribute_9  attribute_10  ...  \
0       0.0986       0.1539       0.1601       0.3109        0.2111  ...   
1       0.2583       0.2156       0.3481       0.3337        0.2872  ...   
2       0.2280       0.2431       0.3771       0.5598        0.6194  ...   
3       0.0368       0.1098       0.1276       0.0598        0.1264  ...   
4       0.0649       0.1209       0.2467       0.3564        0.4459  ...   

   attribute_52  attribute_53  attribute_54  attribute_55  attribute_56  \
0        0.0027        0.0065        0.0159        0.0072        0.0167   
1        0.0084        0.0089        0.0048        0.0094        0.0191   
2        0.0232        0.0166        0.0095        0.0180        0.0244   
3        0.0121        0.0036        0.0150        0.0085        0.0073   
4        0.0031        0.0054        0.0105        0.0110        0.0015   

   attribute_57  attribute_58  attribute_59  attribute_60  Class  
0        0.0180        0.0084        0.0090        0.0032   Rock  
1        0.0140        0.0049        0.0052        0.0044   Rock  
2        0.0316        0.0164        0.0095        0.0078   Rock  
3        0.0050        0.0044        0.0040        0.0117   Rock  
4        0.0072        0.0048        0.0107        0.0094   Rock  

[5 rows x 61 columns]

In [3]:
#Data types
df.dtypes

attribute_1     float64
attribute_2     float64
attribute_3     float64
attribute_4     float64
attribute_5     float64
                 ...   
attribute_57    float64
attribute_58    float64
attribute_59    float64
attribute_60    float64
Class            object
Length: 61, dtype: object

**Data Description**  

This dataset classifies sonar signals which bounced off a metal cylinder and a roughly cylindrical rock.   

**Features/Labels and their datatypes**

The features are a set of 60 floating point numbers in the range 0.0 to 1.0. The label associated with each record contains "Rock" if the object in which was tested was a rock and "Mine" if it was a metal cylinder. Given that we have two labels, this means is a binary classification problem.

**Source**  
[UCI Machine Leaning Repository](https://archive.ics.uci.edu/ml/datasets/Connectionist+Bench+(Sonar,+Mines+vs.+Rocks))

### Data preparation

[Return to top](#Building-a-Multilayer-Neural-Network-to-Classify-Sonar-Signals)  
[Go to Import and data description](#Import-and-data-description)  

In [4]:
#Create a dataframe that holds the input FEATURES
X_input = df.iloc[:, :-1]       

#Create a second dataframe to hold the target LABELS
Y_label = df['Class'].values

In [7]:
#Reshape the labels to make them neural network-compatible using one-hot-encoding
labelencoder_Y = LabelEncoder() 
Y_label = labelencoder_Y.fit_transform(Y_label)
Y_label = Y_label.reshape([208, 1])

### Building the multilayer model

>For the purpose of testing, 4 models were build to compare the differences between different hyperparameter configurations.

[Return to top](#Building-a-Multilayer-Neural-Network-to-Classify-Sonar-Signals)  
[Go to Import and data description](#Import-and-data-description)  
[Go to Data preparation](#Data-preparation)  

In [8]:
#Use Keras Sequential model to build the multilayer model 
model = Sequential()                                     
model.add(Dense(300,input_dim=60, activation = 'relu')) #input layer
model.add(Dense(200,  activation = 'relu'))             #hidden layer 1
model.add(Dense(100,  activation = 'relu'))             #hidden layer 2
model.add(Dense(1,  activation = 'sigmoid'))            #output layer

In [9]:
#Use Keras Sequential model to build the multilayer model
model1 = Sequential()                                     
model1.add(Dense(24,input_dim=60, activation = 'relu')) #input layer
model1.add(Dense(12,  activation = 'relu'))             #hidden layer 1
model1.add(Dense(6,  activation = 'relu'))              #hidden layer 2
model1.add(Dense(3,  activation = 'relu'))              #hidden layer 3
model1.add(Dense(2,  activation = 'relu'))              #hidden layer 4
model1.add(Dense(1,  activation = 'sigmoid'))           #output layer

In [10]:
#Use Keras Sequential model to build the multilayer model
model2 = Sequential()                                     
model2.add(Dense(60,input_dim=60, activation = 'relu')) #input layer
model2.add(Dense(50,  activation = 'relu'))             #hidden layer 1
model2.add(Dense(40,  activation = 'relu'))             #hidden layer 2
model2.add(Dense(30,  activation = 'relu'))             #hidden layer 3
model2.add(Dense(20,  activation = 'relu'))             #hidden layer 4
model2.add(Dense(1,  activation = 'sigmoid'))           #output layer

In [11]:
#Use Keras Sequential model to build the multilayer model
model3 = Sequential()                                     
model3.add(Dense(1000,input_dim=60, activation = 'relu')) #input layer
model3.add(Dense(500,  activation = 'relu'))              #hidden layer 1
model3.add(Dense(1,  activation = 'sigmoid'))             #output layer

In [12]:
#Use Keras Sequential model to build the multilayer model
model4 = Sequential()                                     
model4.add(Dense(1000,input_dim=60, activation = 'relu')) #input layer
model4.add(Dropout(.20))                                  #Randomly deactivate 20% of the neurons in the input layer
model4.add(Dense(500,  activation = 'relu'))              #hidden layer 1
model4.add(Dense(1,  activation = 'sigmoid'))             #output layer

**What is happening in these lines of code**

* **_Sequential_** is a Keras model that facilitates adding layers where each layer has exactly one input tensor
* **_Dense_** is the type of layer that will be added.
* **_mymodel.add(Dense())_** is read as follows:
    * **_units (first parameter within Dense)_** = Number of neurons in the layer (refer to the table below for more details)
    * **_input_dim_** = Number of features **(60 features)
    * **_activation_** = Activation function used in each layer
        * ReLU - Used for multilayer neural network problems
        * Sigmoid - Used in the last layer given this is binary classifier
    
**Hyperparameters used:**  

|                         	|                               model                               	|                               model1                              	|                               model2                              	|                               model3                              	|                               model4                              	|
|-------------------------	|:-----------------------------------------------------------------:	|:-----------------------------------------------------------------:	|:-----------------------------------------------------------------:	|:-----------------------------------------------------------------:	|:-----------------------------------------------------------------:	|
|      **Optimizers**     	|                                Adam                               	|                                Adam                                	|                                Adam                               	|                                Adam                               	|                                Adam                               	|
| **Activation Function** 	| ReLU (for initial and hidden layers)<br>Sigmoid (for last layer)  	| ReLU (for initial and hidden layers)<br>Sigmoid (for last layer)  	| ReLU (for initial and hidden layers)<br>Sigmoid (for last layer)  	| ReLU (for initial and hidden layers)<br>Sigmoid (for last layer)  	| ReLU (for initial and hidden layers)<br>Sigmoid (for last layer)  	|
| **Hidden layers**       	|                                 2                                 	|                                 4                                 	|                                 5                                 	|                                 1                                 	|                                 1                                 	|
| **Units/Neurons**       	|                       **300**-200-100-**1**                       	|                      **24**-12-6-3-2-**1**                      	    |                 **60**-50-40-30-20-**1**                        	|                         **1000**-500-**1**                        	|               **1000**-20% Dropout-500-**1**                        	|
| **Epochs**              	|                                 30                                  	|                                 300                                  	|                                 30                                  	|                                 30                                    |                                 30                                    |

### Training configuration

[Return to top](#Building-a-Multilayer-Neural-Network-to-Classify-Sonar-Signals)  
[Go to Import and data description](#Import-and-data-description)  
[Go to Data preparation](#Data-preparation)  
[Go to Building the multilayer model](#Building-the-multilayer-model)  

In [13]:
#Inspect model configuration
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 300)               18300     
_________________________________________________________________
dense_1 (Dense)              (None, 200)               60200     
_________________________________________________________________
dense_2 (Dense)              (None, 100)               20100     
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 101       
Total params: 98,701
Trainable params: 98,701
Non-trainable params: 0
_________________________________________________________________


In [14]:
#Inspect model1 configuration
model1.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_4 (Dense)              (None, 24)                1464      
_________________________________________________________________
dense_5 (Dense)              (None, 12)                300       
_________________________________________________________________
dense_6 (Dense)              (None, 6)                 78        
_________________________________________________________________
dense_7 (Dense)              (None, 3)                 21        
_________________________________________________________________
dense_8 (Dense)              (None, 2)                 8         
_________________________________________________________________
dense_9 (Dense)              (None, 1)                 3         
Total params: 1,874
Trainable params: 1,874
Non-trainable params: 0
____________________________________________________

In [15]:
#Inspect model2 configuration
model2.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_10 (Dense)             (None, 60)                3660      
_________________________________________________________________
dense_11 (Dense)             (None, 50)                3050      
_________________________________________________________________
dense_12 (Dense)             (None, 40)                2040      
_________________________________________________________________
dense_13 (Dense)             (None, 30)                1230      
_________________________________________________________________
dense_14 (Dense)             (None, 20)                620       
_________________________________________________________________
dense_15 (Dense)             (None, 1)                 21        
Total params: 10,621
Trainable params: 10,621
Non-trainable params: 0
__________________________________________________

In [16]:
#Inspect model3 configuration
model3.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_16 (Dense)             (None, 1000)              61000     
_________________________________________________________________
dense_17 (Dense)             (None, 500)               500500    
_________________________________________________________________
dense_18 (Dense)             (None, 1)                 501       
Total params: 562,001
Trainable params: 562,001
Non-trainable params: 0
_________________________________________________________________


In [17]:
#Inspect model4 configuration
model4.summary()

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_19 (Dense)             (None, 1000)              61000     
_________________________________________________________________
dropout (Dropout)            (None, 1000)              0         
_________________________________________________________________
dense_20 (Dense)             (None, 500)               500500    
_________________________________________________________________
dense_21 (Dense)             (None, 1)                 501       
Total params: 562,001
Trainable params: 562,001
Non-trainable params: 0
_________________________________________________________________


#### Model Inspection Summary
Parameters to learn across layers:  

|                	| **model** 	| **model1** 	| **model2** 	| **model3** 	| **model4** 	|
|----------------	|:---------:	|:----------:	|:----------:	|:----------:	|:----------:	|
| **Parameters** 	|   98,701  	|     1874    	|   10,621  	|   562,001  	|   562,001  	|

[Go to Final remarks](#Final-remarks)  

In [18]:
#Provide the training parameters using compile method
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

In [19]:
#Provide the training parameters using compile method
model1.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

In [20]:
#Provide the training parameters using compile method
model2.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

In [21]:
#Provide the training parameters using compile method
model3.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

In [22]:
#Provide the training parameters using compile method
model4.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

**What is happening in these lines of code**

* **_mymodel.compile_** is read as follows:
     * **optimizer** = **Adam** is the one used for the different tests
     * **loss** = **'binary_crossentropy'** for binary classfiers. Calculates how often predictions equal labels.
     * **metrics** = **'accuracy'** displays information such as the number of epochs, the training loss, the training accuracy, the test loss, and the test accuracy during the training process.

### Model training and evaluation

[Return to top](#Building-a-Multilayer-Neural-Network-to-Classify-Sonar-Signals)  
[Go to Import and data description](#Import-and-data-description)  
[Go to Data preparation](#Data-preparation)  
[Go to Building the multilayer model](#Building-the-multilayer-model)  
[Go to Training configuration](#Training-configuration)  

In [23]:
#Train model
model.fit(X_input, Y_label, epochs=30)

Train on 208 samples
Epoch 1/30
208/208 [==============================] - 0s 2ms/sample - loss: 0.6716 - accuracy: 0.5577
Epoch 2/30
208/208 [==============================] - 0s 120us/sample - loss: 0.6185 - accuracy: 0.7308
Epoch 3/30
208/208 [==============================] - 0s 134us/sample - loss: 0.5759 - accuracy: 0.7067
Epoch 4/30
208/208 [==============================] - 0s 101us/sample - loss: 0.5139 - accuracy: 0.7548
Epoch 5/30
208/208 [==============================] - 0s 110us/sample - loss: 0.4639 - accuracy: 0.8221
Epoch 6/30
208/208 [==============================] - 0s 107us/sample - loss: 0.4139 - accuracy: 0.8317
Epoch 7/30
208/208 [==============================] - 0s 134us/sample - loss: 0.3767 - accuracy: 0.8269
Epoch 8/30
208/208 [==============================] - 0s 127us/sample - loss: 0.3444 - accuracy: 0.8510
Epoch 9/30
208/208 [==============================] - 0s 129us/sample - loss: 0.3243 - accuracy: 0.8750
Epoch 10/30
208/208 [========================

In [24]:
#Evaluate model
model.evaluate(X_input, Y_label)

208/208 [==============================] - 0s 393us/sample - loss: 0.0499 - accuracy: 0.9856


[0.049904003739356995, 0.9855769]

In [25]:
#Train model1
model1.fit(X_input, Y_label, epochs=300)

Train on 208 samples
Epoch 1/300
208/208 [==============================] - 0s 2ms/sample - loss: 0.6950 - accuracy: 0.4712
Epoch 2/300
208/208 [==============================] - 0s 77us/sample - loss: 0.6931 - accuracy: 0.5000
Epoch 3/300
208/208 [==============================] - 0s 96us/sample - loss: 0.6920 - accuracy: 0.5337
Epoch 4/300
208/208 [==============================] - 0s 101us/sample - loss: 0.6908 - accuracy: 0.6346
Epoch 5/300
208/208 [==============================] - 0s 101us/sample - loss: 0.6894 - accuracy: 0.6875
Epoch 6/300
208/208 [==============================] - 0s 91us/sample - loss: 0.6883 - accuracy: 0.7067
Epoch 7/300
208/208 [==============================] - 0s 96us/sample - loss: 0.6868 - accuracy: 0.7308
Epoch 8/300
208/208 [==============================] - 0s 96us/sample - loss: 0.6855 - accuracy: 0.7308
Epoch 9/300
208/208 [==============================] - 0s 86us/sample - loss: 0.6843 - accuracy: 0.7308
Epoch 10/300
208/208 [====================

208/208 [==============================] - 0s 77us/sample - loss: 0.2115 - accuracy: 0.9760
Epoch 156/300
208/208 [==============================] - 0s 82us/sample - loss: 0.2117 - accuracy: 0.9760
Epoch 157/300
208/208 [==============================] - 0s 77us/sample - loss: 0.2122 - accuracy: 0.9760
Epoch 158/300
208/208 [==============================] - 0s 77us/sample - loss: 0.2131 - accuracy: 0.9760
Epoch 159/300
208/208 [==============================] - 0s 72us/sample - loss: 0.2121 - accuracy: 0.9760
Epoch 160/300
208/208 [==============================] - 0s 72us/sample - loss: 0.2078 - accuracy: 0.9760
Epoch 161/300
208/208 [==============================] - 0s 82us/sample - loss: 0.2068 - accuracy: 0.9760
Epoch 162/300
208/208 [==============================] - 0s 91us/sample - loss: 0.2048 - accuracy: 0.9760
Epoch 163/300
208/208 [==============================] - 0s 82us/sample - loss: 0.2043 - accuracy: 0.9760
Epoch 164/300
208/208 [==============================] - 0s 

In [26]:
#Evaluate model1
model1.evaluate(X_input, Y_label)

208/208 [==============================] - 0s 432us/sample - loss: 0.1364 - accuracy: 0.9760


[0.13637146360885638, 0.97596157]

In [27]:
#Train the model2
model2.fit(X_input, Y_label, epochs=30)

Train on 208 samples
Epoch 1/30
208/208 [==============================] - 0s 2ms/sample - loss: 0.6994 - accuracy: 0.4663
Epoch 2/30
208/208 [==============================] - 0s 91us/sample - loss: 0.6862 - accuracy: 0.5769
Epoch 3/30
208/208 [==============================] - 0s 101us/sample - loss: 0.6775 - accuracy: 0.6971
Epoch 4/30
208/208 [==============================] - 0s 96us/sample - loss: 0.6645 - accuracy: 0.7115
Epoch 5/30
208/208 [==============================] - 0s 91us/sample - loss: 0.6476 - accuracy: 0.7260
Epoch 6/30
208/208 [==============================] - 0s 101us/sample - loss: 0.6159 - accuracy: 0.7692
Epoch 7/30
208/208 [==============================] - 0s 101us/sample - loss: 0.5699 - accuracy: 0.7740
Epoch 8/30
208/208 [==============================] - 0s 91us/sample - loss: 0.5185 - accuracy: 0.7885
Epoch 9/30
208/208 [==============================] - 0s 101us/sample - loss: 0.5100 - accuracy: 0.7500
Epoch 10/30
208/208 [============================

In [28]:
#Evaluate the model2
model2.evaluate(X_input, Y_label)

208/208 [==============================] - 0s 460us/sample - loss: 0.1244 - accuracy: 0.9567


[0.12436113993708904, 0.9567308]

In [29]:
#Train the model3
model3.fit(X_input, Y_label, epochs=30)

Train on 208 samples
Epoch 1/30
208/208 [==============================] - 0s 2ms/sample - loss: 0.6846 - accuracy: 0.5481
Epoch 2/30
208/208 [==============================] - 0s 245us/sample - loss: 0.5827 - accuracy: 0.7212
Epoch 3/30
208/208 [==============================] - 0s 254us/sample - loss: 0.5090 - accuracy: 0.7356
Epoch 4/30
208/208 [==============================] - 0s 259us/sample - loss: 0.4729 - accuracy: 0.8029
Epoch 5/30
208/208 [==============================] - 0s 240us/sample - loss: 0.4107 - accuracy: 0.8077
Epoch 6/30
208/208 [==============================] - 0s 235us/sample - loss: 0.4123 - accuracy: 0.7933
Epoch 7/30
208/208 [==============================] - 0s 240us/sample - loss: 0.3537 - accuracy: 0.8846
Epoch 8/30
208/208 [==============================] - 0s 249us/sample - loss: 0.3255 - accuracy: 0.8750
Epoch 9/30
208/208 [==============================] - 0s 259us/sample - loss: 0.3054 - accuracy: 0.8894
Epoch 10/30
208/208 [========================

In [30]:
#Evaluate the model3
model3.evaluate(X_input, Y_label)

208/208 [==============================] - 0s 523us/sample - loss: 0.0398 - accuracy: 1.0000


[0.03979095257818699, 1.0]

In [31]:
#Train the model4
model4.fit(X_input, Y_label, epochs=30)

Train on 208 samples
Epoch 1/30
208/208 [==============================] - 0s 2ms/sample - loss: 0.6695 - accuracy: 0.5962
Epoch 2/30
208/208 [==============================] - 0s 273us/sample - loss: 0.6230 - accuracy: 0.6298
Epoch 3/30
208/208 [==============================] - 0s 254us/sample - loss: 0.5545 - accuracy: 0.6779
Epoch 4/30
208/208 [==============================] - 0s 273us/sample - loss: 0.5210 - accuracy: 0.7452
Epoch 5/30
208/208 [==============================] - 0s 278us/sample - loss: 0.4645 - accuracy: 0.7596
Epoch 6/30
208/208 [==============================] - 0s 288us/sample - loss: 0.4440 - accuracy: 0.7692
Epoch 7/30
208/208 [==============================] - 0s 273us/sample - loss: 0.4016 - accuracy: 0.8269
Epoch 8/30
208/208 [==============================] - 0s 288us/sample - loss: 0.3840 - accuracy: 0.7933
Epoch 9/30
208/208 [==============================] - 0s 283us/sample - loss: 0.3672 - accuracy: 0.8317
Epoch 10/30
208/208 [========================

In [32]:
#Evaluate the model4
model4.evaluate(X_input, Y_label)

208/208 [==============================] - 0s 441us/sample - loss: 0.1190 - accuracy: 0.9567


[0.11901242665659922, 0.9567308]

### Final remarks

[Return to top](#Building-a-Multilayer-Neural-Network-to-Classify-Sonar-Signals)  
[Go to Import and data description](#Import-and-data-description)  
[Go to Data preparation](#Data-preparation)  
[Go to Building the multilayer model](#Building-the-multilayer-model)  
[Go to Training configuration](#Training-configuration)  
[Go to Model training and evaluation](#Model-training-and-evaluation)  

<table>
<tl><th>Model Performance </th></tl>
<tl><td>

|              | **model** | **model1** | **model2** | **model3** | **model4** |
|:-------------|:---------:|:----------:|:----------:|:----------:|:----------:|
| **Loss**     |   0.0555  |   0.6908   |   0.1741   |   0.0333   |   0.0558   |
| **Accuracy** |   0.9903  |   0.5336   |   0.9663   |   0.9951   |   0.9903   |

</td></tl> </table>

After testing several configurations of the models, 4 out 5 models obtained an accuracy of above 95%. It's worth noting the underperforming model (**model1**), in its [summary model1](#Model-Inspection-Summary) we could note the number of parameters for this configuration was significantly smaller(1874) compared to its counterparts. Given the model was too simple for the problem at hand, it underfitted the data. 

The number of epochs for this particular problem didn't require many iterations, it was noted that 30 epochs was enough to minimize the loss for most of the examples.

Finally, the 4 models with accuracies above 95%, 3 of them obtained a loss that ranged close to 0 (0.03 - 0.05) (model, model3, model4). Referring back to the [model summary table](#Model-Inspection-Summary), this three presented larger structures which could lead to overfitting. On the other hand, model2 presented less parameters and a loss not to close to 0. For those reasons, model2 is considered the best configuration. 

***

<table>
<tl><th>Configuration - model2<br>(Best performing model) </th></tl>
<tl><td>
    
|                           |       **model2**                                                     	|
|:--------------------		|:---------------------------------------------------------------:		|
|      **Optimizers**     	|                                 Adam                               	|
| **Activation Function** 	| ReLU (for initial and hidden layers)<br>Sigmoid (for last layer)  	|
| **Hidden layers**       	|                                 5                                 	|
| **Units/Neurons**       	|                **60**-50-40-30-20-**1**                            	|
| **Epochs**              	|                                30                                  	|
|**Parameters**             |  10,621                                                               |

</td></tl> </table>

This model will be put to test again to observe if it gets a better (or worse) performance by removing the last two hidden layers (3 and 4). Lastly, it will be modified one last time by adding one extra hidden layer and distribute the number of neurons among the layers as follows: **70**-60-50-40-30-10-**1**

In [33]:
#Use Keras Sequential model to build the multilayer model
model2 = Sequential()                                     
model2.add(Dense(60,input_dim=60, activation = 'relu')) #input layer
model2.add(Dense(50,  activation = 'relu'))             #hidden layer 1
model2.add(Dense(40,  activation = 'relu'))             #hidden layer 2
#model2.add(Dense(30,  activation = 'relu'))             #remove hidden layer 3
#model2.add(Dense(20,  activation = 'relu'))             #remove hidden layer 4
model2.add(Dense(1,  activation = 'sigmoid'))           #output layer

In [34]:
#Inspect model2 configuration
model2.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_22 (Dense)             (None, 60)                3660      
_________________________________________________________________
dense_23 (Dense)             (None, 50)                3050      
_________________________________________________________________
dense_24 (Dense)             (None, 40)                2040      
_________________________________________________________________
dense_25 (Dense)             (None, 1)                 41        
Total params: 8,791
Trainable params: 8,791
Non-trainable params: 0
_________________________________________________________________


In [35]:
#Provide the training parameters using compile method
model2.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

In [36]:
#Train the model2
model2.fit(X_input, Y_label, epochs=30)

Train on 208 samples
Epoch 1/30
208/208 [==============================] - 0s 2ms/sample - loss: 0.6951 - accuracy: 0.4712
Epoch 2/30
208/208 [==============================] - 0s 91us/sample - loss: 0.6695 - accuracy: 0.5769
Epoch 3/30
208/208 [==============================] - 0s 101us/sample - loss: 0.6513 - accuracy: 0.6394
Epoch 4/30
208/208 [==============================] - 0s 96us/sample - loss: 0.6295 - accuracy: 0.7019
Epoch 5/30
208/208 [==============================] - 0s 86us/sample - loss: 0.6043 - accuracy: 0.7308
Epoch 6/30
208/208 [==============================] - 0s 91us/sample - loss: 0.5779 - accuracy: 0.7356
Epoch 7/30
208/208 [==============================] - 0s 91us/sample - loss: 0.5454 - accuracy: 0.7596
Epoch 8/30
208/208 [==============================] - 0s 91us/sample - loss: 0.5101 - accuracy: 0.7885
Epoch 9/30
208/208 [==============================] - 0s 91us/sample - loss: 0.4784 - accuracy: 0.7981
Epoch 10/30
208/208 [==============================]

In [37]:
#Evaluate the model2
model2.evaluate(X_input, Y_label)

208/208 [==============================] - 0s 393us/sample - loss: 0.2095 - accuracy: 0.9375


[0.20950550213456154, 0.9375]

In [38]:
#Use Keras Sequential model to build the multilayer model. Change neurons as follows: 70-60-50-40-30-10-1
model2 = Sequential()                                     
model2.add(Dense(70,input_dim=60, activation = 'relu')) #input layer
model2.add(Dense(60,  activation = 'relu'))             #hidden layer 1
model2.add(Dense(50,  activation = 'relu'))             #hidden layer 2
model2.add(Dense(40,  activation = 'relu'))             #hidden layer 3
model2.add(Dense(30,  activation = 'relu'))             #hidden layer 4
model2.add(Dense(10,  activation = 'relu'))              #add hidden layer 5
model2.add(Dense(1,  activation = 'sigmoid'))           #output layer

In [39]:
#Inspect model2 configuration
model2.summary()

Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_26 (Dense)             (None, 70)                4270      
_________________________________________________________________
dense_27 (Dense)             (None, 60)                4260      
_________________________________________________________________
dense_28 (Dense)             (None, 50)                3050      
_________________________________________________________________
dense_29 (Dense)             (None, 40)                2040      
_________________________________________________________________
dense_30 (Dense)             (None, 30)                1230      
_________________________________________________________________
dense_31 (Dense)             (None, 10)                310       
_________________________________________________________________
dense_32 (Dense)             (None, 1)                

In [40]:
#Provide the training parameters using compile method
model2.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

In [41]:
#Train the model2
model2.fit(X_input, Y_label, epochs=30)

Train on 208 samples
Epoch 1/30
208/208 [==============================] - 0s 2ms/sample - loss: 0.6905 - accuracy: 0.5337
Epoch 2/30
208/208 [==============================] - 0s 91us/sample - loss: 0.6798 - accuracy: 0.6058
Epoch 3/30
208/208 [==============================] - 0s 96us/sample - loss: 0.6652 - accuracy: 0.6923
Epoch 4/30
208/208 [==============================] - 0s 101us/sample - loss: 0.6438 - accuracy: 0.7067
Epoch 5/30
208/208 [==============================] - 0s 96us/sample - loss: 0.6117 - accuracy: 0.7452
Epoch 6/30
208/208 [==============================] - 0s 101us/sample - loss: 0.5710 - accuracy: 0.7837
Epoch 7/30
208/208 [==============================] - 0s 110us/sample - loss: 0.5266 - accuracy: 0.7692
Epoch 8/30
208/208 [==============================] - 0s 110us/sample - loss: 0.4897 - accuracy: 0.8125
Epoch 9/30
208/208 [==============================] - 0s 91us/sample - loss: 0.4437 - accuracy: 0.8462
Epoch 10/30
208/208 [============================

In [42]:
#Evaluate the model2
model2.evaluate(X_input, Y_label)

208/208 [==============================] - 0s 494us/sample - loss: 0.1030 - accuracy: 0.9856


[0.10300513070363265, 0.9855769]

**When removing the layers**

* The parameters were lowered to 8,791
* The accuracy was slightly lowered to 0.9375
* The loss increased to 0.20950550213456154

**When adding an extra layer and chaging the number of neurons**

* The parameters increased to 15,171
* The accuracy increased to 0.9855769
* The loss was lowered to 0.10300513070363265

While we can get a decent performance with less parameters, is below our 95% accuracy benchmark. On the other hand, with slightly more parameters, there is a small improvement in the results. This only means that the model most likely will perform best with nothing below its original configuration of 4 hidden layers and a neurons distribution **60**-50-40-30-20-**1**

[Return to top](#Building-a-Multilayer-Neural-Network-to-Classify-Sonar-Signals)  
[Go to Import and data description](#Import-and-data-description)  
[Go to Data preparation](#Data-preparation)  
[Go to Building the multilayer model](#Building-the-multilayer-model)  
[Go to Training configuration](#Training-configuration)  
[Go to Model training and evaluation](#Model-training-and-evaluation)  
[Go to Final remarks](#Final-remarks)  